source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.1:
+  //
+  // scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+  //
+  // Although schemes are case-insensitive, the canonical form is lowercase and documents that
+  // specify schemes must do so with lowercase letters. An implementation should accept uppercase
+  // letters as equivalent to lowercase in scheme names (e.g., allow "HTTP" as well as "http") for
+  // the sake of robustness but should only produce lowercase scheme names for consistency.
+  //
+  // The validation mode controls whether uppercase letters are permitted.
+  //
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  auto character_it = value_string_view.begin();
+
+  // The first character must be an ALPHA
+  auto valid_first_character = (*character_it >= 'a' && *character_it <= 'z') ||
+                               (*character_it >= 'A' && *character_it <= 'Z');
+  if (!valid_first_character) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  for (++character_it; character_it != value_string_view.end(); ++character_it) {
+    if (!testChar(kSchemeHeaderCharTable, *character_it)) {
+      return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+    }
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateStatusHeader(const StatusPseudoHeaderValidationMode& mode,
+                                      const HeaderString& value) {
+  //
+  // This is based on RFC 7231, https://datatracker.ietf.org/doc/html/rfc7231#section-6,
+  // describing the list of response status codes and the list of registered response status codes,
+  // https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml.
+  //
+  static const absl::node_hash_set<std::uint32_t> kOfficialStatusCodes = {
+      100, 102, 103, 200, 201, 202, 203, 204, 205, 206, 207, 208, 226, 300, 301, 302,
+      303, 304, 305, 306, 307, 308, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409,
+      410, 411, 412, 413, 414, 415, 416, 417, 418, 421, 422, 423, 424, 425, 426, 428,
+      429, 431, 451, 500, 501, 502, 503, 504, 505, 506, 507, 508, 510, 511,
+  };
+  static const uint32_t kMinimumResponseStatusCode = 100;
+  static const uint32_t kMaximumResponseStatusCode = 599;
+
+  const auto& value_string_view = value.getStringView();
+
+  auto buffer_start = value_string_view.data();
+  auto buffer_end = buffer_start + value_string_view.size();
+
+  // Convert the status to an integer.
+  std::uint32_t status_value{};
+  auto result = std::from_chars(buffer_start, buffer_end, status_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidStatus};
```

Comment:
does Envoy currently reject unregistered status codes?  does the RFC suggest unassigned ones are illegal?

Reply:
Right now we just check that status is a number between 100 and 999. We have relaxed the default HV to do the same.